File: app/src/main/java/igrek/songbook/gui/treelist/TreeItemAdapter.java
package igrek.todotree.gui.treelist;

import android.content.Context;
import android.graphics.Typeface;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import igrek.todotree.R;
import igrek.todotree.gui.GUIListener;
import igrek.todotree.logic.datatree.TreeItem;
import igrek.todotree.system.output.Output;

public class TreeItemAdapter extends ArrayAdapter<TreeItem> {

    Context context;
    List<TreeItem> dataSource;
    List<Integer> selections = null;
    GUIListener guiListener;
    TreeListView listView;

    HashMap<Integer, View> itemViews;
    HashMap<Integer, Integer> itemHeights;

    View convertView = null;
    ViewGroup parent = null;

    public TreeItemAdapter(Context context, List<TreeItem> dataSource, GUIListener guiListener, TreeListView listView) {
        super(context, 0, new ArrayList<TreeItem>());
        this.context = context;
        if (dataSource == null) dataSource = new ArrayList<>();
        this.dataSource = dataSource;
        this.guiListener = guiListener;
        this.listView = listView;
        itemViews = new HashMap<>();
        itemHeights = new HashMap<>();
    }

    public void setDataSource(List<TreeItem> dataSource) {
        this.dataSource = dataSource;
        itemViews = new HashMap<>();
        itemHeights = new HashMap<>();
        notifyDataSetChanged();
    }

    public TreeItem getItem(int position) {
        return dataSource.get(position);
    }

    public void setSelections(List<Integer> selections) {
        this.selections = selections;
    }

    public Integer getViewHeight(int position, ViewGroup listView) {
        if (position >= dataSource.size()) return null;
        if(itemHeights.containsKey(position)) {
            return itemHeights.get(position);
        }
        if (!itemViews.containsKey(position)){
//            itemView = getView(position, convertView, parent);
//            itemView.measure(0, 0);
//            Output.log("no view in map View, position: "+ position + ", new H: " + itemView.getHeight() + ", mh: " + itemView.getMeasuredHeight() + "mhi: " + itemView.getMeasuredHeightAndState());
            Output.log("no view stored in adapter: " + position);
            return null;
        }else {
            View itemView = itemViews.get(position);
            return itemView.getHeight();
        }
    }

    public HashMap<Integer, Integer> getItemHeights() {
        return itemHeights;
    }

    public View getStoredView(int position){
        if (position >= dataSource.size()) return null;
        if (!itemViews.containsKey(position)) return null;
        return itemViews.get(position);
    }

    @Override
    public int getCount() {
        return dataSource.size() + 1;
    }

    @Override
    public long getItemId(int position) {
        if (position < 0) return -1;
        if (position >= dataSource.size()) return -1;
        return (long) position;
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {

        this.convertView = convertView;
        this.parent = parent;

        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        if (position == dataSource.size()) {
            //plusik
            View itemPlus = inflater.inflate(R.layout.item_plus, parent, false);

            ImageButton plusButton = (ImageButton) itemPlus.findViewById(R.id.buttonAddNewItem);
            plusButton.setFocusableInTouchMode(false);
            plusButton.setFocusable(false);
            plusButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    guiListener.onAddItemClicked();
                }
            });

            return itemPlus;
        } else {
            final View itemView = inflater.inflate(R.layout.tree_item, parent, false);
            final TreeItem item = dataSource.get(position);

            //zawartość tekstowa elementu
            TextView textView = (TextView) itemView.findViewById(R.id.tvItemContent);
            StringBuilder contentBuilder = new StringBuilder(item.getContent());
            if (!item.isEmpty()) {
                contentBuilder.append(" [");
                contentBuilder.append(item.size());
                contentBuilder.append("]");
                textView.setTypeface(null, Typeface.BOLD);
            } else {
                textView.setTypeface(null, Typeface.NORMAL);
            }
            textView.setText(contentBuilder.toString());

            //edycja elementu
            ImageButton editButton = (ImageButton) itemView.findViewById(R.id.buttonItemEdit);
            editButton.setFocusableInTouchMode(false);
            editButton.setFocusable(false);
            if (selections == null && !item.isEmpty()) {
                editButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        guiListener.onItemEditClicked(position, item);
                    }
                });
            } else {
                editButton.setVisibility(View.GONE);
            }

            //wejście w element
            ImageButton goIntoButton = (ImageButton) itemView.findViewById(R.id.buttonItemGoInto);
            goIntoButton.setFocusableInTouchMode(false);
            goIntoButton.setFocusable(false);
            if (selections == null && item.isEmpty()) {
                goIntoButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        guiListener.onItemGoIntoClicked(position, item);
                    }
                });
            } else {
                goIntoButton.setVisibility(View.GONE);
            }

            //usuwanie elementu
            ImageButton removeButton = (ImageButton) itemView.findViewById(R.id.buttonItemRemove);
            removeButton.setFocusableInTouchMode(false);
            removeButton.setFocusable(false);
            removeButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    guiListener.onItemRemoveClicked(position, item);
                }
            });

            //przesuwanie
            final ImageButton moveButton = (ImageButton) itemView.findViewById(R.id.buttonItemMove);
            moveButton.setFocusableInTouchMode(false);
            moveButton.setFocusable(false);
            if (selections == null) {
                moveButton.setOnTouchListener(new View.OnTouchListener() {
                    @Override
                    public boolean onTouch(View v, MotionEvent event) {
                        switch (event.getAction()) {
                            case MotionEvent.ACTION_DOWN:
                                listView.onItemMoveButtonPressed(position, item, itemView, event.getX(), event.getY() + moveButton.getTop());
                                return false;
                            case MotionEvent.ACTION_UP:
                                listView.onItemMoveButtonReleased(position, item, itemView, event.getX(), event.getY() + moveButton.getTop());
                                return true;
                        }
                        return false;
                    }
                });
                moveButton.setOnLongClickListener(new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View v) {
                        return listView.onItemMoveLongPressed(position, item);
                    }
                });
            } else {
                moveButton.setVisibility(View.INVISIBLE);
                moveButton.setLayoutParams(new RelativeLayout.LayoutParams(0, 0));
            }

            //dodawanie nowego elementu
            ImageButton addButton = (ImageButton) itemView.findViewById(R.id.buttonItemAddHere);
            addButton.setFocusableInTouchMode(false);
            addButton.setFocusable(false);
            if (selections == null) {
                addButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        guiListener.onAddItemClicked(position);
                    }
                });
            } else {
                addButton.setVisibility(View.GONE);
            }

            //checkbox do zaznaczania wielu elementów
            CheckBox cbItemSelected = (CheckBox) itemView.findViewById(R.id.cbItemSelected);
            cbItemSelected.setFocusableInTouchMode(false);
            cbItemSelected.setFocusable(false);

            if (selections == null) {
                cbItemSelected.setVisibility(View.GONE);
            } else {
                cbItemSelected.setVisibility(View.VISIBLE);
                if (selections.contains(position)) {
                    cbItemSelected.setChecked(true);
                } else {
                    cbItemSelected.setChecked(false);
                }
                cbItemSelected.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                    @Override
                    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                        guiListener.onSelectedClicked(position, item, isChecked);
                    }
                });
            }

            //zapisanie rozmiaru widoku
            itemViews.put(position, itemView);
            itemView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
                @Override
                public void onGlobalLayout() {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                        itemView.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                    }
                    itemHeights.put(position, itemView.getHeight());
                    //Output.log("stored height = " + itemView.getHeight() + ", pos: " + position);
                }
            });

            return itemView;
        }
    }

}

File: app/src/main/java/igrek/songbook/playlist/treelist/TreeItemAdapter.kt
package igrek.songbook.playlist.treelist

import android.content.Context
import android.graphics.Rect
import android.text.SpannableString
import android.text.style.UnderlineSpan
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.TouchDelegate
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.RelativeLayout

import java.util.ArrayList

import igrek.todotree.R
import igrek.todotree.commands.ItemEditorCommand
import igrek.todotree.commands.ItemSelectionCommand
import igrek.todotree.domain.treeitem.AbstractTreeItem
import igrek.todotree.domain.treeitem.LinkTreeItem
import igrek.todotree.ui.errorcheck.SafeClickListener

internal class TreeItemAdapter(context: Context, dataSource: List<AbstractTreeItem>?, private val listView: TreeListView) : ArrayAdapter<AbstractTreeItem>(context, 0, ArrayList<E>()) {

    var items: List<AbstractTreeItem>? = null
        private set
    private var selections: Set<Int>? = null // selected indexes
    private val storedViews: SparseArray<View>
    private val inflater: LayoutInflater

    init {
        var dataSource = dataSource
        if (dataSource == null)
            dataSource = ArrayList<AbstractTreeItem>()
        this.items = dataSource
        storedViews = SparseArray()
        inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    }

    fun setDataSource(dataSource: List<AbstractTreeItem>) {
        this.items = dataSource
        storedViews.clear()
        notifyDataSetChanged()
    }

    override fun getItem(position: Int): AbstractTreeItem? {
        return items!![position]
    }

    fun setSelections(selections: Set<Int>) {
        this.selections = selections
    }

    fun getStoredView(position: Int): View? {
        return if (position >= items!!.size) null else storedViews.get(position)
    }

    override fun getCount(): Int {
        return items!!.size + 1
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    override fun getItemId(position: Int): Long {
        if (position < 0)
            return -1
        return if (position >= items!!.size) -1 else position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        return if (position == items!!.size) {
            getAddItemView(position, parent)
        } else {
            getItemView(position, parent)
        }
    }

    private fun getItemView(position: Int, parent: ViewGroup): View {
        // get from cache
        if (storedViews.get(position) != null)
            return storedViews.get(position)

        val item = items!![position]
        val itemView: View
        if (!item.isEmpty()) {
            itemView = getParentItemView(item, position, parent)
        } else {
            itemView = getSingleItemView(item, position, parent)
        }

        // store view
        storedViews.put(position, itemView)

        return itemView
    }

    private fun getSingleItemView(item: AbstractTreeItem, position: Int, parent: ViewGroup): View {
        val itemView = inflater.inflate(R.layout.tree_item_single, parent, false)

        //zawartość tekstowa elementu
        val textView = itemView.findViewById(R.id.tvItemContent)
        textView.setText(item.getDisplayName())

        // link
        if (item is LinkTreeItem) {
            val content = SpannableString(item.getDisplayName())
            content.setSpan(UnderlineSpan(), 0, content.length(), 0)
            textView.setText(content)
        }

        //przesuwanie
        val moveButton = itemView.findViewById(R.id.buttonItemMove)
        moveButton.setFocusableInTouchMode(false)
        moveButton.setFocusable(false)
        moveButton.setClickable(false)
        increaseTouchArea(moveButton, 20)
        if (selections == null) {
            moveButton.setOnTouchListener({ v, event ->
                event.setSource(777) // from moveButton
                when (event.getAction()) {
                    MotionEvent.ACTION_DOWN -> {
                        listView.reorder
                                .onItemMoveButtonPressed(position, item, itemView, event.getX(), event
                                        .getY() + moveButton.getTop())
                        return@moveButton.setOnTouchListener false
                    }
                    MotionEvent.ACTION_MOVE -> return@moveButton.setOnTouchListener
                    false
                            MotionEvent . ACTION_UP -> {
                        listView.reorder
                                .onItemMoveButtonReleased(position, item, itemView, event.getX(), event
                                        .getY() + moveButton.getTop())
                        return@moveButton.setOnTouchListener true
                    }
                }
                false
            })
        } else {
            moveButton.setVisibility(View.INVISIBLE)
            moveButton.setLayoutParams(RelativeLayout.LayoutParams(0, 0))
        }

        //checkbox do zaznaczania wielu elementów
        val cbItemSelected = itemView.findViewById(R.id.cbItemSelected)
        cbItemSelected.setFocusableInTouchMode(false)
        cbItemSelected.setFocusable(false)

        if (selections != null) {
            cbItemSelected.setVisibility(View.VISIBLE)
            if (selections!!.contains(position)) {
                cbItemSelected.setChecked(true)
            } else {
                cbItemSelected.setChecked(false)
            }
            cbItemSelected.setOnCheckedChangeListener({ buttonView, isChecked ->
                ItemSelectionCommand()
                        .selectedItemClicked(position, isChecked)
            })
        }

        itemView.setOnTouchListener(TreeItemTouchListener(listView, position))

        return itemView
    }

    private fun getParentItemView(item: AbstractTreeItem, position: Int, parent: ViewGroup): View {
        val itemView = inflater.inflate(R.layout.tree_item_parent, parent, false)

        //zawartość tekstowa elementu
        val textView = itemView.findViewById(R.id.tvItemContent)
        textView.setText(item.getDisplayName())

        // ilość potomków
        val tvItemChildSize = itemView.findViewById(R.id.tvItemChildSize)
        val contentBuilder = "[" + item.size() + "]"
        tvItemChildSize.setText(contentBuilder)

        //edycja elementu
        val editButton = itemView.findViewById(R.id.buttonItemEdit)
        editButton.setFocusableInTouchMode(false)
        editButton.setFocusable(false)
        editButton.setClickable(true)
        increaseTouchArea(editButton, 20)
        if (selections == null && !item.isEmpty()) {
            editButton.setOnClickListener(object : SafeClickListener() {
                fun onClick() {
                    ItemEditorCommand().itemEditClicked(item)
                }
            })
        } else {
            editButton.setVisibility(View.GONE)
        }

        //przesuwanie
        val moveButton = itemView.findViewById(R.id.buttonItemMove)
        moveButton.setFocusableInTouchMode(false)
        moveButton.setFocusable(false)
        moveButton.setClickable(false)
        increaseTouchArea(moveButton, 20)
        if (selections == null) {
            moveButton.setOnTouchListener({ v, event ->
                event.setSource(777) // from moveButton
                when (event.getAction()) {
                    MotionEvent.ACTION_DOWN -> {
                        listView.reorder
                                .onItemMoveButtonPressed(position, item, itemView, event.getX(), event
                                        .getY() + moveButton.getTop())
                        return@moveButton.setOnTouchListener false
                    }
                    MotionEvent.ACTION_MOVE -> return@moveButton.setOnTouchListener
                    false
                            MotionEvent . ACTION_UP -> {
                        listView.reorder
                                .onItemMoveButtonReleased(position, item, itemView, event.getX(), event
                                        .getY() + moveButton.getTop())
                        return@moveButton.setOnTouchListener true
                    }
                }
                false
            })
        } else {
            moveButton.setVisibility(View.INVISIBLE)
            moveButton.setLayoutParams(RelativeLayout.LayoutParams(0, 0))
        }

        //checkbox do zaznaczania wielu elementów
        val cbItemSelected = itemView.findViewById(R.id.cbItemSelected)
        cbItemSelected.setFocusableInTouchMode(false)
        cbItemSelected.setFocusable(false)

        if (selections != null) {
            cbItemSelected.setVisibility(View.VISIBLE)
            if (selections!!.contains(position)) {
                cbItemSelected.setChecked(true)
            } else {
                cbItemSelected.setChecked(false)
            }
            cbItemSelected.setOnCheckedChangeListener({ buttonView, isChecked ->
                ItemSelectionCommand()
                        .selectedItemClicked(position, isChecked)
            })
        }

        itemView.setOnTouchListener(TreeItemTouchListener(listView, position))

        return itemView
    }

    private fun getAddItemView(position: Int, parent: ViewGroup): View {
        // plus
        val itemPlus = inflater.inflate(R.layout.item_plus, parent, false)

        val plusButton = itemPlus.findViewById(R.id.buttonAddNewItem)
        plusButton.setFocusableInTouchMode(false)
        plusButton.setFocusable(false)
        plusButton.setOnClickListener(object : SafeClickListener() {
            fun onClick() {
                ItemEditorCommand().addItemClicked()
            }
        })
        // redirect long click to tree list view
        plusButton.setLongClickable(true)
        plusButton.setOnLongClickListener({ v -> listView.onItemLongClick(null, null, position, 0) })

        return itemPlus
    }

    private fun increaseTouchArea(component: View, sidePx: Int) {
        val parent = component.parent as View  // button: the view you want to enlarge hit area
        parent.post {
            val rect = Rect()
            component.getHitRect(rect)
            rect.top -= sidePx    // increase top hit area
            rect.left -= sidePx   // increase left hit area
            rect.bottom += sidePx // increase bottom hit area
            rect.right += sidePx  // increase right hit area
            parent.touchDelegate = TouchDelegate(rect, component)
        }
    }
}